[WPE] WPE Platform: add support for pointer lock

siwei-li opened the pull request:
#### 141c440e3b1ae8d55acaf1d0f92eb6d892891365
<pre>
[WPE] WPE Platform: add support for pointer lock
<a href="https://bugs.webkit.org/show_bug.cgi?id=265648">https://bugs.webkit.org/show_bug.cgi?id=265648</a>

Reviewed by Carlos Garcia Campos.

Pointer lock requests from WebKit are sent to the platform, where we check
if pointer constraints and relative pointer protocols are available in
the Wayland implementation, then the pointer is locked and a corresponding
WPEEvent emitted.

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRequestPointerLock):
(webkitWebViewDidLosePointerLock):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
(WKWPE::View::requestPointerLock):
(WKWPE::View::didLosePointerLock):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::requestPointerLock):
(WKWPE::ViewPlatform::didLosePointerLock):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_lock_pointer):
(wpe_view_unlock_pointer):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandGetPointerConstraints):
(wpeDisplayWaylandGetRelativePointerManager):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandDispose):
(wpeViewWaylandUnlockPointer):
(wpe_view_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.h:
(WPE::WaylandSeat::pointerModifiers const):
(WPE::WaylandSeat::pointerCoords const):

Canonical link: <a href="https://commits.webkit.org/282558@main">https://commits.webkit.org/282558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c763f1d929dcd4b932a932698e914c1913926dd9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15567 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66992 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13575 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50014 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50762 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9366 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39328 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54538 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31445 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36018 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12451 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57562 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68687 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6917 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11829 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58076 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6949 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58278 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5770 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9597 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38147 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39227 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40338 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38969 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3037,6 +3037,10 @@ void webkitWebViewRequestPointerLock(WebKitWebView* webView)
     webkitWebViewBaseRequestPointerLock(WEBKIT_WEB_VIEW_BASE(webView));
 #endif
+
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
```
```suggestion
#if PLATFORM(WPE)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3049,8 +3053,12 @@ void webkitWebViewDidLosePointerLock(WebKitWebView* webView)
 #endif
-}
+
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
```
```suggestion
#if PLATFORM(WPE)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
```diff
@@ -96,6 +96,10 @@ class View : public API::ObjectImpl<API::Object::Type::View> {
+#if ENABLE(POINTER_LOCK)
+    virtual void requestPointerLock() = 0;
+    virtual void didLosePointerLock() = 0;
+#endif
```
Instead of pure virtual, make them virtual with empty implementation for the old api.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```diff
@@ -573,6 +573,24 @@ void ViewPlatform::setCursor(const WebCore::Cursor& cursor)
+void ViewPlatform::requestPointerLock()
+{
+    if (!m_wpeView)
+        return;
```
This can't be nullptr. We have an early return in setCursor() that was copied when we used the same class, but now that we have WPEWebViewPlatform, it can't be nullptr.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```diff
@@ -573,6 +573,24 @@ void ViewPlatform::setCursor(const WebCore::Cursor& cursor)
+void ViewPlatform::didLosePointerLock()
+{
+    if (!m_wpeView)
+        return;
```
Same here

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -755,6 +755,44 @@ void wpe_view_unmap(WPEView* view)
+ * wpe_view_lock_pointer:
+ * @view: a #WPEView
+ *
+ * Disable the movements of the pointer in @view, locking it to one position.
```
This comment is confusing, we are not locking the pointer to one position, but to a particular area. I would explain here that while the pointer is locked, mouse move events are relative instead of absolute.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -755,6 +755,44 @@ void wpe_view_unmap(WPEView* view)
+    if (viewClass->lock_pointer)
+        return viewClass->lock_pointer(view);
+
+    return false;
```
```suggestion
    return FALSE;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -755,6 +755,44 @@ void wpe_view_unmap(WPEView* view)
+    if (viewClass->unlock_pointer)
+        return viewClass->unlock_pointer(view);
+
+    return false;
```
```suggestion
    return FALSE;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+        double x = wl_fixed_to_double(fixedX);
+        double y = wl_fixed_to_double(fixedY);
+        double dX = wl_fixed_to_double(deltaX);
+        double dY = wl_fixed_to_double(deltaY);
```
These are not absolution position and delta, these are accelerated and unaccelerated delta, I think we are only interested in unaccelerated so we can ignore the other.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+        double dX = wl_fixed_to_double(deltaX);
+        double dY = wl_fixed_to_double(deltaY);
+
+        auto* event = wpe_event_pointer_move_new(WPE_EVENT_POINTER_MOVE, view, WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(0), x, y, dX, dY);
```
I think we need to save the pointer position when the pointer was locked to be used here as x, y

carlosgcampos replied:
And I didn't notice before (sorry), but we should also save the modifiers and pass them here instead of 0

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+    auto* pointerConstraints = wpeDisplayWaylandGetPointerConstraints(display);
+    auto* relativePointerManager = wpeDisplayWaylandGetRelativePointerManager(display);
+    if (!pointerConstraints || !relativePointerManager)
+        return false;
```
```suggestion
        return FALSE;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+        return false;
+
+    auto* wlPointer = wpeDisplayWaylandGetWlPointer(display);
+    auto* relativePointer = zwp_relative_pointer_manager_v1_get_relative_pointer(relativePointerManager, wlPointer);
```
We need to save this to destroy it on unlock.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+
+    zwp_relative_pointer_v1_add_listener(relativePointer, &relativePointerListener, view);
+    auto* wlSurface = wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(view));
+    zwp_pointer_constraints_v1_lock_pointer(pointerConstraints, wlSurface, wlPointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
```
This returns a `struct zwp_locked_pointer_v1*` that also needs to be destroyed on unlock.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+    auto* wlSurface = wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(view));
+    zwp_pointer_constraints_v1_lock_pointer(pointerConstraints, wlSurface, wlPointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
+
+    return true;
```
```suggestion
    return TRUE;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+static gboolean wpeViewWaylandUnlockPointer(WPEView* view)
+{
+    auto* display = WPE_DISPLAY_WAYLAND(wpe_view_get_display(view));
+    auto* wlPointer = wpeDisplayWaylandGetWlPointer(display);
```
We should check if we are locked and return FALSE otherwise. We can probably check if we have a relative pointer and locked pointer.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+{
+    auto* display = WPE_DISPLAY_WAYLAND(wpe_view_get_display(view));
+    auto* wlPointer = wpeDisplayWaylandGetWlPointer(display);
+    g_clear_pointer(&wlPointer, zwp_locked_pointer_v1_destroy);
```
You are releasing here the pointer not the locked pointer object.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +401,48 @@ static const struct wl_buffer_listener cursorBufferListener = {
+    auto* wlPointer = wpeDisplayWaylandGetWlPointer(display);
+    g_clear_pointer(&wlPointer, zwp_locked_pointer_v1_destroy);
+
+    return true;
```
```suggestion
    return TRUE;
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/tempCodeRunnerFile.cpp:
```diff
@@ -0,0 +1 @@
+gtk_window_maximize
```
What's this new file?

siwei-li replied:
Sry I think it got accidentally added when updating my local repo

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -58,6 +60,10 @@ struct _WPEViewWaylandPrivate {
+
+    struct zwp_relative_pointer_v1* relativePointer;
+    struct zwp_locked_pointer_v1* lockedPointer;
+    std::pair<double, double> pointerCoords { 0, 0 };
```
```suggestion
    std::pair<double, double> savedPointerCoords { 0, 0 };
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -755,6 +755,46 @@ void wpe_view_unmap(WPEView* view)
+ * area; while the pointer is locked, mouse events are relative instead of
+ * absolute motions.
+ *
+ * Returns: %TRUE if view has support for pointer lock, or %FALSE otherwise
```
We are also returning FALSE when already locked, so I would say something like %TRUE if pointer is locked, simialr to the unlock one

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -399,6 +405,55 @@ static const struct wl_buffer_listener cursorBufferListener = {
+static gboolean wpeViewWaylandUnlockPointer(WPEView* view)
+{
+    auto* priv = WPE_VIEW_WAYLAND(view)->priv;
+    g_clear_pointer(&priv->lockedPointer, zwp_locked_pointer_v1_destroy);
+    g_clear_pointer(&priv->relativePointer, zwp_relative_pointer_v1_destroy);
```
These should be done in dispose too, in case the view is destroyed while pointer is locked.

siwei-li committed:
[WPE] WPE Platform: add support for pointer lock
https://bugs.webkit.org/show_bug.cgi?id=265648

Reviewed by Carlos Garcia Campos.

Pointer lock requests from WebKit are sent to the platform, where we check
if pointer constraints and relative pointer protocols are available in
the Wayland implementation, then the pointer is locked and a corresponding
WPEEvent emitted.

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRequestPointerLock):
(webkitWebViewDidLosePointerLock):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
(WKWPE::View::requestPointerLock):
(WKWPE::View::didLosePointerLock):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::requestPointerLock):
(WKWPE::ViewPlatform::didLosePointerLock):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_lock_pointer):
(wpe_view_unlock_pointer):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandGetPointerConstraints):
(wpeDisplayWaylandGetRelativePointerManager):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandDispose):
(wpeViewWaylandUnlockPointer):
(wpe_view_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.h:
(WPE::WaylandSeat::pointerModifiers const):
(WPE::WaylandSeat::pointerCoords const):

Canonical link: https://commits.webkit.org/282558@main
